samcli/lib/build/app_builder.py

```diff
@@ -441,8 +441,14 @@ def _build_layer(
                     # Only set to this value if specified workflow is makefile
                     # which will result in config language as provided
                     build_runtime = compatible_runtimes[0]
+                image = None
+                if self._build_images is not None:
+                    if layer_name in self._build_images:
+                        image = self._build_images[layer_name]
+                    elif None in self._build_images:
+                        image = self._build_images[None]
                 self._build_function_on_container(
```

Comment:
I am slightly confused on this. Which I think is mainly due to the nesting. 

A couple questions:
1) Can `self._build_images` be `None`? If so, is there a way we can safely default this to an empty dict instead? 
2) Why are we checking if the `layer_name` is in `self._build_images`?
3) I am not familiar with the `build_images` property and how it gets set. Could you describe the contents?

Something to think about for the future:
When you are writing deeply nested code, consider ways you can simplify it. There are usually ways to write the same logic with less control flows or splitting the nesting into simpler private functions. So in this case, if we could default `self._build_images` to an empty dict in `__init__` then we safe ourselves the need to do `None` checks (we do this throughout the code base for this reason). 

When acting on dictionaries in python, try to avoid using `[]` to access elements and instead use `.get()` method. The `.get()` method does the checks you want for you. So you can simplify both inner if statements by doing `image = self._build_images.get(layer_name)`. By default, this will return `None` if the element (`layer_name`) is not in the dict.

I think I actually answered my above questions while writing the last part but will leave the in the comment to help explain my thought process/questions I would ask.

Reply:
1. Yes it is defaulted to be none. Type hinting doesn't allow any parameter defaults to be {}, so no.
2. self._build_images has function names and layer names as keys and image uri's as value.
3. Details in https://github.com/aws/aws-sam-cli/pull/2725, basically we take the command line argument and make a dict that stores the URI's and their related logicalId's

Reply:
+1 on using get(), this is following the previous pr's pattern, will change both to get()

Reply:
And we can set self._build_images to {} during init, I agree totally.